tinygrad/runtime/ops_metal.py

```diff
@@ -65,19 +68,33 @@ def metal_src_to_library(device:MetalDevice, src:str) -> objc_instance:
   return library
 
 class MetalCompiler(Compiler):
-  def __init__(self, dev:Optional[MetalDevice]=None):
-    self.dev = dev
-    super().__init__("compile_metal_xcode" if self.dev is None else "compile_metal")
+  def __init__(self):
+    if not getenv("METAL_XCODE", 0): self.cgs = compiler.MTLCodeGenServiceCreate(b"tinygrad")
+    super().__init__("compile_metal_xcode" if getenv("METAL_XCODE", 0) else "compile_metal")
+  def __reduce__(self): return (MetalCompiler,()) # to create new instance for each multiprocessing fork
   def compile(self, src:str) -> bytes:
-    if self.dev is None:
+    if self.cgs is None:
       # NOTE: if you run llvm-dis on "air" you can see the llvm bytecode
       air = subprocess.check_output(['xcrun', '-sdk', 'macosx', 'metal', '-x', 'metal', '-c', '-', '-o', '-'], input=src.encode('utf-8'))
       lib = subprocess.check_output(['xcrun', '-sdk', 'macosx', 'metallib', '-', '-o', '-'], input=air)
     else:
-      library = metal_src_to_library(self.dev, src)
-      library_contents = msg(library, "libraryDataContents", restype=objc_instance)
-      lib = ctypes.string_at(msg(library_contents, "bytes"), cast(int, msg(library_contents, "length", restype=ctypes.c_ulong)))
-    assert lib[:4] == b"MTLB", "Invalid Metal library. Using conda? Corrupt XCode?"
+      lib = None
+      err: Optional[Exception] = CompileError("MTLCodeGenServiceBuildRequest returned without calling the callback")
+      @ctypes.CFUNCTYPE(None, ctypes.c_void_p, ctypes.c_int32, ctypes.c_void_p, ctypes.c_size_t, ctypes.c_char_p)
+      def callback(blockptr, error, dataPtr, dataLen, errorMessage):
+        nonlocal lib, err
+        if error != 0:
+          err = CompileError(errorMessage.decode())
+          return
+        reply = bytes(to_mv(dataPtr, dataLen))
+        offset = struct.unpack('<Q', reply[8:16])[0]
+        lib, err = reply[offset:], None
+      params = "-fmetal-math-mode=safe -fmetal-math-fp32-functions=precise -std=metal3.0 --driver-mode=metal -x metal"
```

Comment:
It looks like "-fmetal-math-mode=safe -fmetal-math-fp32-functions=precise" aren't available on old versions of macos, i'll find a way to fix this and then mark it ready to review again